src/mem/threadalloc.h

```diff
@@ -243,21 +265,24 @@ namespace snmalloc
    * The simple check that this is the global placeholder is inlined, the rest
    * of it is only hit in a very unusual case and so should go off the fast
    * path.
+   * The second component of the return indicates if this TLS is being torndown.
    */
-  SNMALLOC_SLOW_PATH inline void* init_thread_allocator()
+  SNMALLOC_FAST_PATH void* init_thread_allocator(std::function<void*(void*)>& f)
   {
     auto*& local_alloc = ThreadAlloc::get_reference();
-    if (local_alloc != get_GlobalPlaceHolder())
+    // If someone reuses a noncachable call, then we can end up here
+    // with an already initialised allocator. Could either error
+    // to say stop doing this, or just give them the initialised version.
+    if (local_alloc == get_GlobalPlaceHolder())
     {
-      // If someone reuses a noncachable call, then we can end up here.
-      // The allocator has already been initialised. Could either error
-      // to say stop doing this, or just give them the initialised version.
-      return local_alloc;
+      local_alloc = current_alloc_pool()->acquire();
     }
-    local_alloc = current_alloc_pool()->acquire();
-    SNMALLOC_ASSERT(local_alloc != get_GlobalPlaceHolder());
-    ThreadAlloc::register_cleanup();
-    return local_alloc;
+    // Check if we have already run the destructor for the TLS.  If so, then
+    // we need the caller to handle this specially as we are in teardown.
```

Comment:
I’m not sure what “the caller” means in this context. Is it the caller of register_cleanup, aka the current function? Because I don’t see what the caller of init_thread_allocator needs to do

Reply:
Stale comment from before I added the `std::function`.